frontend/apps/reader/modules/readerfooter.lua

```diff
@@ -591,7 +591,7 @@ function ReaderFooter:init()
     end
 
     self.text_container = RightContainer:new{
-        dimen = Geom:new{ w = 0, h = self.height },
+        dimen = Geom:new{ x = 0, y = 0, w = 0, h = self.height },
```

Comment:
It feels like I've seen that before - either you already did it, or you tried to and we may have discussed it :)

Anyway, this still feel strange to me, having to put these x/y where they are not needed and make not the sense they'd have if explicitely written - because we're not putting that stuff at x=0 y=0. And I'll probably forget to write them :)
If they are needed (ie. to make 4 slots in the table object and avoid Lua having to resize that table), it feels it should be internal to `Geom:new/init()`.

Reply:
> Geom:new/init()

That would be too late. The point is to make the initial table the "right" size, and that's the table being fed to new ;).

Note that some of these are quite likely overly pedantic, I just blindly filled 'em up whenever I encountered one.

Reply:
pedantic/blindly vs. clarity/expliciteness: no x/y has always meant it's a not-(yet-)positionned dimen.
Is this really just to avoid a Lua table resize ? Is that really necessary vs. the unexpiciteness? Are you sure a table resize, and how costly, happens when adding 2 items to a initial 2-items table?

Reply:
> no x/y has always meant it's a not-(yet-)positionned dimen

To me it often looks like "uh, I dunno how this really works, so I'm going to leave that alone ;p" ^^.

If you find that (0, 0) is confusing, we could use a sentinel value instead (I dunno, -1 or 0xFFFF or something). I would tend to prefer (0, 0), though (if only because it feels like zero-initializing in C/C++).

> Are you sure a table resize, and how costly, happens when adding 2 items to a initial 2-items table?

Well, it's a `realloc` that's easily avoidable. I don't get the readability concerns myself (as I mentioned above, I find *notù specifying the coords *more* confusing and wafflish), but they should trump everything else, so, RFC.

Reply:
Can it be instead of
https://github.com/koreader/koreader/blob/e0ed04a1a147162c0c03e5864df2e62e7a5b9ba2/frontend/ui/geometry.lua#L30-L42
```Lua
local Geom = {}

function Geom:new(o)
    local new_o = {
        x = o.x or 0, -- left origin
        y = o.y or 0, -- top origin
        w = o.w or 0, -- width
        h = o.h or 0, -- height
    }
    setmetatable(new_o, self)
    self.__index = self
    return new_o
end

```
And do not pass default zeroes. Initial table in the caller will be gc'ed. No table resizing.

Reply:
Interesting thought, but that would cause GC churn instead, which I'm not convinced would be a great trade ;).

On the other hand, for completely empty Geoms,

```diff
diff --git a/frontend/ui/geometry.lua b/frontend/ui/geometry.lua
index 7c5d27197..792806bca 100644
--- a/frontend/ui/geometry.lua
+++ b/frontend/ui/geometry.lua
@@ -35,7 +35,12 @@ local Geom = {
 }
 
 function Geom:new(o)
-    if not o then o = {} end
+    if not o then
+        o = {
+            x = 0, y = 0,
+            w = 0, h = 0,
+        }
+    end
     setmetatable(o, self)
     self.__index = self
     return o
```

makes sense. Doesn't really address (most of) what we're talking, about though ;p.

Reply:
The purpose is to avoid passing `x = 0, y = 0`.

Reply:
> The purpose is to avoid passing `x = 0, y = 0`.

I get that, but it ends up making *everything* worse from a performance standpoint, compared to only *sometimes* maybe needing a realloc or not depending on the constructor table ;).

Reply:
No other good idea :/
(A bad one, requiring a function call instead: `function Geom.getNewGeom(w, h, x, y) ...` to which we could not provide x/y and that would build and return one table here.)

> making everything worse from a performance standpoint

How much would that be vs. all the other things happening when Geom objects are created, = when widgets are created: there's probably a lot of other objects created/resized, + repaints/filling blitbuffer/refreshs so that the realloc is negligible :)

(But there's not much of my widgets that get their Geom stuff fixed in the current state of this PR :) so may be I did that right (I don't think so :)) or your eyes and mind were not that strong, and you will allow some uncomplete ones to go by if not in hot codepaths - and I would compromise on letting you add them to bits you consider hot-widget-paths, but not making that a hard rule :))

Reply:
> How much would that be vs. all the other things happening

Probably noise. But my point was more that it would somewhat gratuitously make *every* caller "worse" vs. the current status quo or only *some* (and even for those, that's a mild maybe: there's definitely going to be inheritance lookups, but actual writes? That's somewhat rarer).

Reply:
There, left classes that don't actually care about dimensions alone, massively reducing the amount of changes ;).